Migrate utilities in CSS files imported into layers

thecrypticace opened the pull request:
When a stylesheet is imported with `@import “…” layer(utilities)` that means that all classes in that stylesheet and any of its imported stylesheets become candidates for `@utility` conversion.

Doing this correctly requires us to place `@utility` rules into separate stylesheets (usually) and replicate the import tree without layers as `@utility` MUST be root-level. If a file consists of only utilities we won't create a separate file for it and instead place the `@utility` rules in the same stylesheet.

Been doing a LOT of pairing with @RobinMalfait on this one but I think this is finally ready to be looked at

RobinMalfait committed:
Compare filesystem snapshots

Here we introduce a new  `dumpFiles` helper that can be used to list all files matching a given glob and their content. This simplifies some assertions and ensures files are not accidentally added or deleted.

thecrypticace committed:
Cleanup

thecrypticace committed:
Don’t emit `undefined` into a stylesheet

Sometimes a node may not have a `before` raws prop

thecrypticace committed:
Track parent/child relationships across stylesheets

CSS files may be imported and we’ll need to know what files import other files in order to unlock more advanced transformations

thecrypticace committed:
Extract `Stylesheet` class

The purpose of the `Stylesheet` class is to collect and hold relevant information about CSS files that can be used for more advanced transformations.

Additionally this unlocks two things:
- Formatting only happens after all transformations have completed
- Files are only written once at the end — which will simplify the introduction of a `--dry-run` option in the future

thecrypticace committed:
Migrate utilities in layered stylesheets

When a stylesheet is imported with `@import “…” layer(utilities)` that means that all classes in that stylesheet and any of its imported stylesheets become candidates for `@utility` conversion

philipp-spiess commented on packages/@tailwindcss-upgrade/src/migrate.ts:
```diff
@@ -122,3 +128,211 @@ export async function analyze(stylesheets: Stylesheet[]) {
+
+    let utilitySheet = await Stylesheet.fromRoot(
+      utilities,
+      sheet.file.replace(/\.css$/, '.utilities.css'),
```
A file with this name can already exist. If it does, we should error with something useful

thecrypticace replied:
This has been addressed

philipp-spiess reviewed (commented):
This makes sense and is pretty straight-forward when reviewed by commits. Two things to consider though:

1. Let's add a test for this scenario:
   ```css
   --- ./a.css ---
   @import './c.css' layer(utilities);
   --- ./b.css ---
   @import './c.css';
   --- ./c.css ---
   .foo {
     color: red;
   }
   ```
   
   I would expect `c.css` to _stills split into two files_ in this case with one not having the `.foo` class migrated to `@utility` or so. What do you think?

2. We create new files now with `[filename].utilities.css`. Let's check if this already exited prior the migration though and how an error accordingly.

philipp-spiess commented on integrations/upgrade/index.test.ts:
```diff
@@ -485,3 +485,64 @@ test(
+
+    expect(await fs.dumpFiles('./src/**/*.css')).toMatchInlineSnapshot(`
+      "
+      --- ./src/a.1.css ---
+      @utility foo-from-a {
+        color: red;
+      }
```
Hm so this is now upgraded even for the import in `root.3.css` which is wrong. Question is how likely this is to be an issue though? 🤔

RobinMalfait replied:
Haha yeah this test is definitely wrong, @thecrypticace and I are working on it but wanted to push a test to start from. Going to convert this PR to draft until we are confident again.

thecrypticace replied:
We now warn on this case and won't touch the files that can't be converted.

adamwathan commented:
In v3 we never preserved the actual `@layer` rule for `components` and `utilities`, everything was just ultimately inlined no matter what, just in a specific location. Does that simplify anything?

Like in v3, this input:

```
// main.css
@import "my-crap.css" layer(utilities);

// my-crap.css
[potato] {
  background: brown;
}
```

...produced this output:

```
[potato] {
  background: brown;
}
```

...with no `@layer utilities` or anything present at all. So in terms of a codemod that preserves the same behavior shouldn't _everything_ imported into a utilities or components layer be rendered at the top-level anyways?

thecrypticace commented:
@adamwathan Nah, I don't think it does. While v3 did produce that behavior we're still aiming to put things into the appropriate layer when migrating to v4 which means that `[foo]` should still be imported into a utility layer if it follows `@import "tailwind/utilities"` for example.

thecrypticace committed:
Detect conflicting imports for potential utility files

Co-authored-by: Robin Malfait <malfait.robin@gmail.com>

thecrypticace committed:
Place `@utility` rules into separate stylesheets as needed

Now that we convert layered, imported rules into `@utility` where needed we need to hoist these rules outside of their imported layer since `@utility` MUST be root-level.

adamwathan commented:
> @adamwathan Nah, I don't think it does. While v3 did produce that behavior we're still aiming to put things into the appropriate layer when migrating to v4 which means that `[foo]` should still be imported into a utility layer if it follows `@import "tailwind/utilities"` for example.

Do you mean this?

```css
@import "tailwindcss/utilities";

[foo] {
  background: red;
}
```

If so I don't think I agree, should remain unlayered, but also don't think that's what this PR is about anyways if I understand correctly so don't mean to derail.

I guess what I'm trying to think through in my head is what we would do if we were upgrading a project that had this in it by hand with a human touch:

```css
/* main.css */
@import "my-components.css" layer(components);
@import "my-utilities.css" layer(utilities);

/* my-components.css */
.btn {
  background: blue;
  color: white;
  padding: 16px;
}

[potato] {
  background: brown;
}

/* my-utilities.css */
.tab-1 {
  tab-size: 1;
}

[foo] {
  background: red;
}
```

What would we have written instead if this was a v4 project?

thecrypticace commented:
Well… I would probably hoist _all_ utilities to either the root or to their own file. But I think we also want to operate under principle of least surprise — if someone's placed their utilities in a deeper import tree then there's probably a structural reason they did that.

But — we could just collect them all and spit them out in their own file at the end. It would make some of this process much simpler.

Unfortunately we still have the multiple "roots" problem to deal with. If you import some into a utilities layer and some not then regardless the problem still exists for some paths to a specific file.

Aside: I wouldn't've placed utilities in imports like _in the first place_ because I didn't even realize that worked until you pointed it out so I ultimately would never have run into this problem at all.

re the `[foo]` case:

We're talking about imported files only here _but_ in any case how is this different from when we have to wrap imports?

We turn this:
```
@import "tailwindcss/utilities";
@import "./foo.css";
```

into this (was already the case before this PR):
```
@import "tailwindcss/utilities";
@import "./foo.css" layer(utilities);
```

which, if inlined, is the same as this:

```
@import "tailwindcss/utilities";
[foo] {
  background: red;
}
```

becoming this — which I think we also do cc @RobinMalfait:

```
@import "tailwindcss/utilities";
@layer utilities {
  [foo] {
    background: red;
  }
}
```

RobinMalfait commented:
```css
@import "tailwindcss/utilities";
[foo] {
  background: red;
}
```

This would stay as-is (we don't wrap the rules _after_ the last tailwind directive (or import) into an `@layer`). But I think we still do add the `layer(utilities)` after the last tailwind directive.

So I think this is still true:

Input:
```css
@import "tailwindcss/utilities";
@import "./foo.css";
```

Output:
```css
@import "tailwindcss/utilities" layer(utilities);
@import "./foo.css" layer(utilities);
```
